### ISSUE TITLE
The APL content not scaling dynamically to fit the display size

### ISSUE THREAD

**Issue by yacula** [alexa/alexa-smart-screen-sdk#18]
**IMPORTANT**: Before you create an issue, please take a look at our [Issue Reporting Guide](https://github.com/alexa/avs-device-sdk/wiki/Issue-Reporting-Guide).

## Issue:
I have an issue with new Smart Screen SDK. I've noticed that the APL rendering is not scaling to the size of the display. I've followed the documentation of Alexa.Display.Window (https://developer.amazon.com/en-US/docs/alexa/alexa-voice-service/display-window.html) and I see there is an option to specify window template size configuration to CONTINUOUS, which I assume should fix the issue. However when I set minimum and maximum values for scaling the SDK throws an error:
`GuiConfigHelpers.ts:260 Uncaught TypeError: Cannot read property 'value' of undefined`

When I look into the code, I see there is no implementation for this feature in the SDK:
```javascript
if (windowTemplate) {
        let windowSizeConfiguration : IDisplayWindowSize;
        for (let ws of windowTemplate.configuration.sizes) {
            if (ws.id === windowSizeId) {
                windowSizeConfiguration = ws;
                break;
            }
        }
        if (windowSizeConfiguration) {
            const windowSize : IDisplayWindowDiscreteSize =
                windowSizeConfiguration as IDisplayWindowDiscreteSize;
            pixelWidth = windowSize.value.value.width;
            pixelHeight = windowSize.value.value.height;
        }
    }
```

I'm attaching the whole GUI config file for reference:
```json
{
    "gui": {
        "appConfig": {
            "description": "Smart Screen Landscape Large Device",
            "mode": "HUB",
            "emulateDisplayDimensions": false,
            "scaleToFill": true,
            "audioInputInitiator": "TAP",
            "windows": [
                {
                    "id": "smartScreenLandscape",
                    "templateId": "smartScreenLandscape",
                    "sizeConfigurationId": "fullscreen",
                    "interactionMode": "smart_screen"
                }
            ],
            "defaultWindowId": "smartScreenLandscape",
            "deviceKeys": {
                "talkKey": {
                    "code": "KeyA",
                    "keyCode": 65,
                    "key": "a"
                },
                "backKey": {
                    "code": "KeyB",
                    "keyCode": 66,
                    "key": "b"
                },
                "exitKey": {
                    "code": "Escape",
                    "keyCode": 27,
                    "key": "Escape"
                }
            }
        },
        "visualCharacteristics": [
            {
                "type": "AlexaInterface",
                "interface": "Alexa.InteractionMode",
                "version": "1.0",
                "configurations": {
                    "interactionModes": [
                        {
                            "id": "smart_screen",
                            "uiMode": "HUB",
                            "interactionDistance": {
                                "unit": "INCHES",
                                "value": 84
                            },
                            "touch": "SUPPORTED",
                            "keyboard": "UNSUPPORTED",
                            "video": "SUPPORTED"
                        }
                    ]
                }
            },
            {
                "type": "AlexaInterface",
                "interface": "Alexa.Presentation.APL.Video",
                "version": "1.0",
                "configurations": {
                    "video": {
                        "codecs": [
                            "H_264_42",
                            "H_264_41"
                        ]
                    }
                }
            },
            {
                "type": "AlexaInterface",
                "interface": "Alexa.Display.Window",
                "version": "1.0",
                "configurations": {
                    "templates": [
                        {
                            "id": "smartScreenLandscape",
                            "type": "STANDARD",
                            "configuration": {
                                "sizes": [
                                    {
                                        "type": "CONTINUOUS",
                                        "id": "fullscreen",
                                        "minimum": {
                                            "unit": "PIXEL",
                                            "value": {
                                                "width": 1280,
                                                "height": 800
                                            }
                                        },
					                    "maximum": {
                                            "unit": "PIXEL",
                                            "value": {
                                                "width": 2000,
                                                "height": 2000
                                            }
                                        }
                                    }
                                ],
                                "interactionModes": [
                                    "smart_screen"
                                ]
                            }
                        }
                    ]
                }
            },
            {
                "type": "AlexaInterface",
                "interface": "Alexa.Display",
                "version": "1.0",
                "configurations": {
                    "display": {
                        "type": "PIXEL",
                        "touch": [
                            "SINGLE"
                        ],
                        "shape": "RECTANGLE",
                        "dimensions": {
                            "resolution": {
                                "unit": "PIXEL",
                                "value": {
                                    "width": 1280,
                                    "height": 800
                                }
                            },
                            "physicalSize": {
                                "unit": "INCHES",
                                "value": {
                                    "width": 8.6,
                                    "height": 5.4
                                }
                            },
                            "pixelDensity": {
                                "unit": "DPI",
                                "value": 160
                            },
                            "densityIndependentResolution": {
                                "unit": "DP",
                                "value": {
                                    "width": 1280,
                                    "height": 800
                                }
                            }
                        }
                    }
                }
            }
        ]
    }
}
```

## Expected behavior
I assume using this template size configuration:
```json
{
   "type":"CONTINUOUS",
   "id":"fullscreen",
   "minimum":{
      "unit":"PIXEL",
      "value":{
         "width":1280,
         "height":800
      }
   },
   "maximum":{
      "unit":"PIXEL",
      "value":{
         "width":2000,
         "height":2000
      }
   }
}
```
should make the rendered APL content fit the display size within the boundaries provided as minimum and maximum. 

## Observed behavior
When I set minimum and maximum values I get the following error in the browser:
`GuiConfigHelpers.ts:260 Uncaught TypeError: Cannot read property 'value' of undefined`

If I don't specify the template at all the rendered content is not filling the entire screen, and leaves some margins on the sides.

### AVS Device SDK version:
      <1.15>
### Smart Screen SDK version:
      <2.0.1>

### Tell us what hardware you're using:
- [x] Desktop / Laptop
- [ ] Raspberry Pi
- [ ] Other - tell us more:

### Tell us about your OS (Type & version):
- [x] Linux
- [ ] MacOS 
- [ ] Raspbian Stretch
- [ ] Raspbian Jessy
- [ ] Other - tell us more:

**Comment by nguymi91** [comment]
@yacula, the response for this matter is quite messy.

The Alexa.Display.Window API (https://developer.amazon.com/en-US/docs/alexa/alexa-voice-service/display-window.html) is not implicitly tied to APL, it’s just an API that allows a device to report the type of windows it has. We have released CONTINUOUS window types (i.e. windows that can change size based on user input) as reportable types in 1.0.  That in no way means that there’s an APL runtime or APL content built to respond to dynamic window size changes, just allows a user to tell someone that they have a resizable window.

No APL content will currently automatically scale to fill a viewport, because of the whole reported supportedViewports framework (https://developer.amazon.com/en-US/docs/alexa/alexa-presentation-language/apl-select-the-viewport-profiles-your-skill-supports.html#how-different-devices-display-your-apl-content).  So, ALL APL content will currently be scaled to preserve aspect ratio within a non-standard reported viewport, there is no responsive scaling.  This is the case regardless of if a DISCRETE or CONTINUOUS window is reported.

At this moment, we don't support any concept of a dynamically resizing APL window. We're planning on handling this case more gracefully in the future.

For the time being, please switch "CONTINUOUS" to "DISCRETE".

Thank you,
Mike

**Comment by nguymi91** [comment]
Closing this issue for now. Please re-open in the future if needed.

**Comment by yacula** [comment]
Hi @nguymi91 !
Recently I was investigating the issue of responsive rendering for Alexa skills and I came across an interesting case. I've noticed that the weather skill is displayed properly on any aspect ratio (although there are still some issues with explicit portrait orientation). I've updated my skill to APL 1.4 and tried to use AlexaDetail template to render the content, however I still have a black bar on the right side of the screen. I've also noticed the same behaviour for "Tell me a joke" feature. I'm attaching screenshots for reference. Is there currently any way of implementing responsive design to Alexa skills with APL 1.4?
When I design the visual response in the Visual Response Builder in the Alexa console I can set any screen size and the content is displayed correctly. However, when I actually invoke the skill I get black bars. Is this related to the APL library or APLRenderer or can it be somehow forced on the skill implementation level? I would very much appreciate some guidance and I'm sorry if this is not a proper place to ask these questions.

Cheers!

P.S. I cannot actually post a screenshot from my skill (work stuff) but I see the same behaviour as for "Tell me a joke" output.
<img width="811" alt="Screenshot 2020-11-03 at 15 38 37" src="https://user-images.githubusercontent.com/20949308/98000463-ce909000-1dec-11eb-8514-e856b98cdee6.png">
<img width="811" alt="Screenshot 2020-11-03 at 15 39 10" src="https://user-images.githubusercontent.com/20949308/98000472-d2bcad80-1dec-11eb-8163-df951cabd663.png">
<img width="321" alt="Screenshot 2020-11-03 at 15 44 00" src="https://user-images.githubusercontent.com/20949308/98000475-d4867100-1dec-11eb-91b6-39a673484a6c.png">
